### ISSUE TITLE
Support Encrypted Client Hello (formerly known as ESNI)

### ISSUE THREAD

**Issue by Gunni** [caddyserver/caddy#4221]
Please add support for [TLS Encrypted Client Hello](https://datatracker.ietf.org/doc/draft-ietf-tls-esni/)

- Championed by the EFF: https://www.eff.org/deeplinks/2018/09/esni-privacy-protecting-upgrade-https
- Deployed by Cloudflare: https://blog.cloudflare.com/esni/
- Cloudflare's technical details post: https://blog.cloudflare.com/encrypted-sni/
- Supported by Firefox: https://blog.mozilla.org/security/2018/10/18/encrypted-sni-comes-to-firefox-nightly/
- Supported by NSS: https://bugzilla.mozilla.org/show_bug.cgi?id=1495120
- Work in Progress in picotls: h2o/picotls#155

[Previous ticket](https://github.com/caddyserver/caddy/issues/2327) for ESNI was not recreated after being closed, so here it is.

List taken from [the openssl issue](https://github.com/openssl/openssl/issues/7482).

**Comment by francislavoie** [comment]
https://blog.cloudflare.com/encrypted-client-hello/

I don't think ESNI has a future.

And this isn't actionable for the Caddy project until Golang implements it. And there seems to be no plans to unless there's suddenly wider adoption.

**Comment by mholt** [comment]
ECH is its successor though. If Go implements it we should try to support it.

**Comment by francislavoie** [comment]
Oh sorry, I totally misread the issue title. My bad!

**Comment by joshxyzhimself** [comment]
Related

- https://guardianproject.info/2021/11/30/implementing-tls-encrypted-client-hello/
- https://news.ycombinator.com/item?id=29530811

**Comment by diyism** [comment]
Any updates? all of nginx,apache,lighttpd have support ECH, how about caddy?

**Comment by francislavoie** [comment]
@diyism I can't find any evidence that they do support it. Where did you see that? Please share links.

**Comment by diyism** [comment]
> @diyism I can't find any evidence that they do support it. Where did you see that? Please share links.

I've test it in https://defo.ie/ with msedge dev(https://www.thewindowsclub.com/enable-encrypted-client-hello-in-microsoft-edge):

nginx:  https://draft-13.esni.defo.ie:10413
apache2: https://draft-13.esni.defo.ie:11413
both of the upper two pages show "SSL_ECH_STATUS: success" and a green tick mark.
Can it prove they hava a ECH-ready nginx and apache2 ?

**Comment by francislavoie** [comment]
Those are experiments using a custom build of OpenSSL. That's not really "support".

Either way, Caddy doesn't use OpenSSL, it uses Go's stdlib for TLS. So this work doesn't apply to us.

**Comment by mholt** [comment]
Yeah, that would be like us building with a custom fork of the Go standard library to support ECH. We could do that ([Cloudflare has one](https://github.com/cloudflare/go)), but given the sensitive nature of the `crypto/tls` package, I would only recommend that for experimental or low-risk environments for now. (CF is surely using it in production, but we don't know how much of the diff is tailored to just their use cases.)

So I guess we can answer "How about Caddy?" with "Sure, go for it" -- just build it with a forked tls package and it is possible. (I'm uncertain if any code changes to Caddy would need to take place, but since you're building from source already, might as well.)

As a side-note, the ECH draft spec seems to have expired: https://datatracker.ietf.org/doc/draft-ietf-tls-esni/

**Comment by ViViDboarder** [comment]
I think this doesn’t change anything as Caddy’s stance appears to be “when it’s supported in Go stdlib,” which makes total sense to me, but it looks like ECH has legs and is now supported in mainline Firefox: https://blog.mozilla.org/en/products/firefox/encrypted-hello/

**Comment by mholt** [comment]
Yeah; to clarify, we like ECH, and will support it someday assuming Go's standard lib implements it. (Because we're wary about switching to a fork of crypto/tls for our standard distribution of Caddy.)

However, I should note that ECH has limited practical use currently as DNS lookups still reveal hostnames in plaintext most of the time, and packets still have IP addresses that aren't encrypted.

It's just really, really hard to solve the problem people are hoping to solve, and ECH is not the whole solution, and without the whole solution, users are still at risk.

But it's cool I guess.

**Comment by ViViDboarder** [comment]
Firefox also defaults to DNS over TLS as well, so this is another step to close the gaps.

IP addresses is a big gap. Though it’s one that can be mitigated, depending on trade offs one is willing to make. I like to self host, so I run Caddy on my own server with my IP. So my connection is going to be known that way, but others, who may not care about their own end to end control, can run their server behind a CDN, like Cloudflare. In that case, the IP revealed is one that gives an interloper very little information.

I’m glad to have seen more and more progress over the last few years and excited to see what’s to come!

**Comment by dennisjackson** [comment]
@mholt - Just to clarify, ECH is the last part of the solution. There were four places where website names were leaking in plaintext: 
 - In the TLS Server Certificate (fixed by TLS 1.3) 
 - In the DNS lookup (fixed by DNS-over-HTTPS - which has already shipped in Firefox) 
 - In the certificate revocation checking (fixed by OCSP Stapling and CRLite)
 - In the SNI field (fixed by ECH). 
 
On the topic of IPs, they are far less unique than people realise. Cloudflare have already experimented with serving all of their websites from a single IP address [1]. In their measurement study, over 50% of the entire .com, .net, .info and .org namespace (255 million websites) is available from only ~100 IPs. This is obviously not great from a centralisation perspective, but that seems like more a reason for Caddy to support privacy features like ECH, not less!

**Comment by mholt** [comment]
> Firefox also defaults to DNS over TLS as well, so this is another step to close the gaps.

I'm curious how widespread this is; if it's mostly us tech geeks who actually use this feature -- I know, for example, my parents don't benefit from this. I'd still wager the vast majority of DNS lookups are cleartext.

And yeah, IP addresses in _some_ cases don't reveal much, inasmuch as we centralize to a single hosting provider.

>  but that seems like more a reason for Caddy to support privacy features like ECH, not less!

To AGAIN clarify, we like ECH. I just want to make sure that people understand that it is only a part of the solution and the solution is still largely incomplete. Your sites and web browsing experience may already have the whole solution, but (without seeing stats I can only guess) I'd wager that most of the Internet still has a long way to go, so I want to make sure we don't hype up something that isn't real for most people yet.

**Comment by mholt** [comment]
(What I'm saying is that most people don't use Firefox, unfortunately.)

**Comment by Harusakii** [comment]
> (What I'm saying is that most people don't use Firefox, unfortunately. I know other browsers plan to roll it out too, but we're not there yet.)

I think we are there now, Chrome released the update 117.0.5938.150 which now enables ECH on default. Just saying though ^^

**Comment by networkException** [comment]
Where did you see that? Checking out chromium source at the `117.0.5938.150` tag declares `kEncryptedClientHello` as `FEATURE_DISABLED_BY_DEFAULT` in `net/base/features.cc`

**Comment by Harusakii** [comment]
Idk about Chromium but as of https://chromestatus.com/roadmap it is in Origin Trial but still planing to ship in 117 so it will come later, but either way Chrome did enable it by default (you may need to reset your chrome://flags) and then you can check by going to https://www.cloudflare.com/ssl/encrypted-sni to see it enabled

**Comment by alx-alexpark** [comment]
This should be a high priority, as it renders many methods of censorship useless.

**Comment by mohammed90** [comment]
We're dependent on Go standard library to support this function. Watch golang/go#63369. It seems to be going in a positive direction.

**Comment by rugk** [comment]
Firefox now supports ECH, would be great if you could support it, too: https://blog.mozilla.org/en/products/firefox/encrypted-hello/

**Comment by francislavoie** [comment]
@rugk as it's been said multiple times in this thread, there's nothing we (the Caddy team) can do until the Go stdlib has support for ECH.

**Comment by PrivacyOsint2** [comment]
@francislavoie @mholt @mohammed90 @Gunni 

crypto/tls: add ech client support

https://github.com/golang/go/commit/9eeb627f606e713854e117dd4e52af5bcad28b66

**Comment by PrivacyOsint2** [comment]
https://github.com/golang/go/issues/63369

**Comment by Gunni** [comment]
> @francislavoie @mholt @mohammed90 @Gunni 
> 
> They added support for ECH.
> 
> crypto/tls: add ech client support
> 
> https://github.com/golang/go/commit/9eeb627f606e713854e117dd4e52af5bcad28b66

Only client side logic https://github.com/golang/go/issues/63369#issuecomment-2126161985

**Comment by coolaj86** [comment]
Status has been updated to Accepted: https://github.com/golang/go/issues/68500

Release targeted for Go 1.24: https://github.com/golang/go/issues/63369#issuecomment-2083070008

**Comment by mholt** [comment]
As I design this, my current understanding of ECH is, based on the [ESNI draft 22](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#name-overview):

> A client-facing server enables ECH by publishing an ECH configuration, which is an encryption public key and associated metadata. The server must publish this for all the domains it serves via Shared or Split Mode.

So:

- At least one key pair, but possibly others, are generated by the server (or admin -- but let's be real, we're fond of automation here).

On publishing keys:

> This document defines the ECH configuration's format, but delegates DNS publication details to [[RFC9460](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#RFC9460)]. See [[ECH-IN-DNS](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#ECH-IN-DNS)] for specifics about how ECH configurations are advertised in HTTPS records. Other delivery mechanisms are also possible. For example, the client may have the ECH configuration preconfigured.

- The public key is published, usually to DNS records (!) ... but also may be preconfigured into clients (so, kinda like SSH in a way).
- The relevant draft RFC with details for servers is here: https://datatracker.ietf.org/doc/html/draft-ietf-tls-svcb-ech-05#name-server-behavior
- There's also RFC 9460 about the HTTPS RRs: https://www.rfc-editor.org/rfc/rfc9460#https

The Go standard library exposes one property for ECH servers, [`EncryptedClientHelloKeys`](https://pkg.go.dev/crypto/tls#Config.EncryptedClientHelloKeys). The tricky thing is it takes a `Config []byte`, which is the already-serialized form of the ECHConfig struct defined by the ESNI draft spec. There is no standard function that produces this serialization... yet. There is [this package](https://github.com/OmarTariq612/goech) which seems to do this, but it'd be nice if we didn't have new dependencies for this.

Anyway... bringing this all down to our task here to integrate this capability into Caddy:

To make this as easy, and as least error-prone as possible, for Caddy users, I suppose we could expose a config object for ECH in the TLS connection policies that enable ECH. The user would have to either provide keys or have Caddy automate them. Of course, we'll prefer the latter.

We have the unique opportunity to do something really special here, with the [`libdns` packages](https://github.com/libdns) we already use for ACME DNS challenges and things like the [caddy-dynamicdns plugin](https://github.com/mholt/caddy-dynamicdns). With ECH enabled, and DNS provider credentials in the config, we can generate ECH keys, publish them to DNS records, and serve ECH with regularly-rotated keys. i.e. this could be really easy when it comes to user-facing config.

Any thoughts before I take this further? Am I way off base here?

**Comment by christaikobo** [comment]
Apologies in advance if my question sounds stupid here.

The cloudflare implementation seems to have an outer and inner SNI. I'm
wondering if that is compulsory and what plan caddy has regarding this
issue.

**Comment by mholt** [comment]
I got it working!

It's currently very spikey, but I just needed to prove my understanding of things and show that it works. I will be refactoring and refining it, and looking at addressing the security considerations in the RFC as soon as possible.

Config example (will be changing, but should remain about this simple):

![Image](https://github.com/user-attachments/assets/d59db025-8dc1-4517-8b35-4c04a48a2db8)

I'll likely hoist this to a global config setting, since I imagine most users will want a single outer domain that hides all their sites, right?

That config causes Caddy to provision an HTTPS-type DNS record with the `ech` SvcParamKey containing the public key and the outer SNI. Caddy then manages a certificate for the site's domain as usual, and also the outer SNI.

Then in Firefox with DoH (DNS-over-HTTPS) enabled, I loaded the domain name of the site I was trying to access:

![Image](https://github.com/user-attachments/assets/33153de2-23dc-44c9-b594-32eaf5133d76)

And sure enough it loaded. But was it hidden? Wireshark says yes:

![Image](https://github.com/user-attachments/assets/60ab832a-1586-4627-998a-73303669ad70)

Only the outer SNI is visible on the wire.

---

There's still a lot to consider. I have no idea yet if I'm doing it _properly_ other than it seems to be working. I will study the RFC more.

---

@christaikobo 

> The cloudflare implementation seems to have an outer and inner SNI. I'm wondering if that is compulsory and what plan caddy has regarding this issue.

I think it is... at least as much as SNI is "compulsory" without ECH. The outer SNI is needed (?) so a visible handshake can be completed and a certificate served. The inner SNI is the actual hostname being accessed.

> How will it interact with the layer4 addon? More specifically, the SNI mathers?

Not quite sure yet... layer4 will need to be updated to support ECH. The ECH spec defines two modes: "shared" and "split". In shared mode, the server both unwraps the inner SNI, and terminates TLS. In split mode, the server only unwraps the inner SNI, and proxies the "real" TLS handshake to a backend without terminating it. (Pretty cool IMO, although layer4 already supports bespoke TLS-in-TLS tunnels for any clients that happen to support that.)

@devhaozi 

> This package https://github.com/c2FmZQ/ech also can use to generate ECH config.

Thanks a lot for this; actually the package I found doesn't seem to marshal the ECHConfig correctly, but that package does... I likely won't be using either, since the actual borrowed code is only about 15 lines long, and both packages had a nearly identical implementation (which makes sense, given that there's only 1 proper way to do it.)

**Comment by devhaozi** [comment]
> I think it is... at least as much as SNI is "compulsory" without ECH. The outer SNI is needed (?) so a visible handshake can be completed and a certificate served. The inner SNI is the actual hostname being accessed.

From my understanding, outer SNI is required, but it should be customizable to be different from inner SNI. (In [Cloudflare's implementation](https://developers.cloudflare.com/ssl/edge-certificates/ech/) they use `cloudflare-ech.com` as outer SNI)

**Comment by mholt** [comment]
I figured out what was wrong with the package at https://github.com/OmarTariq612/goech -- it marshals an ECHConfig as an ECHConfigList (with 1 ECHConfig in it), and the MarshalBinary for ECHConfigList, I think, improperly encodes a single length at the beginning instead of a length per element.

Anyway, I have made my own implementation that Caddy will use. The HTTPS DNS record takes a base64-encoded ECHConfigList and the Go crypto/tls package takes one or more binary-encoded ECHConfig (paired with the private keys).

I have that all working now.

It seems that a certificate is _not_ required for the outer SNI, which is convenient.

The remaining task is to figure out how to design the Caddy config.

**Comment by christaikobo** [comment]
> It seems that a certificate is _not_ required for the outer SNI, which is convenient.

Interesting, does this mean we will be allowed to use any domain as outer SNI?

**Comment by Gunni** [comment]
What about not using an outer SNI? Is that permissible? Or spoofing any outer name if we aren't serving it anyway?

**Comment by christaikobo** [comment]
> What about not using an outer SNI? Is that permissible? Or spoofing any outer name if we aren't serving it anyway?

Yes this is pretty much what is going through my mind. For large cooperation it makes sense to have an outer SNI in order to hide an inner SNI. For smaller organization and individuals with only 1 or a couple of domains, the effectiveness is drastically reduced. 

For example, hiding `blog.mydomain.com` with `ech.mydomain.com` is pretty much useless if `mydomain.com` is the only domain I have.

**Comment by mholt** [comment]
@christaikobo and @Gunni 

> Interesting, does this mean we will be allowed to use any domain as outer SNI?

> What about not using an outer SNI? Is that permissible? Or spoofing any outer name if we aren't serving it anyway?

No; to clarify, a certificate for the outer/public name is optional, but is recommended. If the server cannot decrypt the client's ECH extension (e.g. old or wrong config presented), then the server can complete the handshake with the outer ClientHello (thus it would need a certificate) so the client can safely get the updated configs to retry (https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#rejected-ech). Basically, the outer SNI must be controlled by (or delegated to) the ECH server for successful, tracking-resilient client retries.

https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#section-8.1.1:

> public_name: The DNS name of the client-facing server, i.e., the entity trusted to update the ECH configuration. This is used to correct misconfigured clients, as described in [Section 6.1.6](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#rejected-ech).
> ...
> The retry mechanism repairs inconsistencies, provided the server is authoritative for the public name. ... Provided the server can present a certificate valid for the public name, the client can safely retry with updated settings, as described in [Section 6.1.6](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#rejected-ech).
> ..
> In order to ensure that the retry mechanism works successfully servers SHOULD ensure that every endpoint which might receive a TLS connection is provisioned with an appropriate certificate for the public name. This is especially important during periods of server reconfiguration when different endpoints might have different configurations.

Inconsistencies are possible as the latest DNS records are not always visible to clients, or whatever other method of publication is being employed; the point is that the client and the server both need to have =exact byte-for-byte the-same ECHConfigs. If they don't, my understanding is that the client can use a retry mechanism to resolve inconsistencies. However, doing this in a way that doesn't defeat the purpose of ECH requires the origin server to provide a certificate for the public name.

See also: https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#name-authenticating-for-the-publ

**In summary:** No, technically, an ECH server does NOT need a certificate for the public name. However, this can make client reliability more brittle, since it will not be able to safely resolve inconsistencies / update its state. Thus, the server SHOULD be authoritative for that public name. Caddy will still manage a certificate for the public name, but maybe there'll be an option to disable that requirement. It can still serve ECH to clients while it gets its first certificate for the public name.

---

> For large cooperation it makes sense to have an outer SNI in order to hide an inner SNI. For smaller organization and individuals with only 1 or a couple of domains, the effectiveness is drastically reduced.
>
> For example, hiding blog.mydomain.com with ech.mydomain.com is pretty much useless if mydomain.com is the only domain I have.

True. A DNS scan can reveal which "hidden" names are using ECH and what their corresponding public names are. The spec also says:

> Usage of this mechanism reveals that a client is connecting to a particular service provider, but does not reveal which server from the anonymity set terminates the connection.
> ...
> ECH is not in itself sufficient to protect the identity of the server. The target domain may also be visible through other channels, such as plaintext client DNS queries or visible server IP addresses. However, DoH [[RFC8484](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#RFC8484)] and DPRIVE [[RFC7858](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#RFC7858)] [[RFC8094](https://www.ietf.org/archive/id/draft-ietf-tls-esni-22.html#RFC8094)] provide mechanisms for clients to conceal DNS lookups from network inspection, and many TLS servers host multiple domains on the same IP address. Private origins may also be deployed behind a common provider, such as a reverse proxy. In such environments, the SNI remains the primary explicit signal used to determine the server's identity.

This is actually one valid criticism of ECH: it tends to favor centralization; i.e. bigcorps like Cloudflare/Google/Amazon hosting all the things, since the more things that are hidden behind a central authority, the more anonymous/ambiguous the visible parts become.

**Comment by christaikobo** [comment]
Thanks for the elaborated explanation, very informative. 

I was wrong to get my hopes a little bit too high. But if caddy decides to ship with the option for disabling the requirement of a public certification, I will play around a bit and see how it affects things. Maybe it's a compromise some of us are willing to make, reduced reliability in exchange for better anonymity.

**Comment by mholt** [comment]
I'm not sure not having a public name affords you "better anonymity," and I'm worried that this might be a common and slippery/dangerous mistake that a lot of people assume. I believe, the reality is, public name does not practically reveal more information than the IP address being connected to.

The spec refers to an **anonymity set** as this:

> Co-located servers with consistent externally visible TLS configurations and behavior, including supported versions and cipher suites and how they respond to incoming client connections, form an anonymity set.

So, if you're on Cloudflare, Google, Amazon, DigitalOcean, etc., then your anonymity set, in theory, could be quite large. If you are self-hosting on your home Internet service and relying on ECH for anonymity, your anonymity set is drastically smaller, regardless of whether a public name is used or not, by the mere fact that a set is comprised of _co-located servers_.

The spec advises making your anonymity set as large as possible:

> With regards to handshake privacy, client-facing server configuration determines the size of the anonymity set. For example, if a client-facing server uses distinct ECHConfig values for each host, then each anonymity set has size k = 1. Client-facing servers SHOULD deploy ECH in such a way so as to maximize the size of the anonymity set where possible. This means client-facing servers should use the same ECHConfig for as many hosts as possible.

(In terms of anonymity, there should be as few ECHConfigs as possible on the Internet; which of course calls up the criticism that ECH is advocating centralization, but I digress.)

A public name does not affect these factors at all, and only improves the reliability of clients being able to connect. So, I don't think specifying a public name in the config affects the anonymity of the setup.

In other words, based on my current limited understanding, using a public name is orthogonal to anonymity. Sure, don't choose a name like `ech-for-hidden-service.com`, but even a random domain name would do fine.

Remember, too, that hidden names are still publicly known (assuming the public Internet & DNS), since the HTTPS-type record that will conventionally be used to publish ECH configs are also public and can be determined with a very simple DNS lookup. So the fact that a name is being hidden is public information regardless. (You could, theoretically, hide a name without everyone knowing it is hidden by publishing ECH configs privately, but that's... obviously not publicly useful.)

So, what about specifying a public name materially reduces anonymity?

**Comment by christaikobo** [comment]
I am thinking about SNI based censorship, especially (but not limited to) a whitelist approach. Masquerading as `cloudflare-ech.com` would get you pass a lot (not all) of these. Most firewall system are not going to conduct targeted and sophisticated investigation, so we don't need absolute anonymity, we just need enough of it to get by. (IMHO the thing about privacy and anonymity is always a trade off, you can never achieve absolute safety)

What you have talked about are undeniably true, in a relatively ideal world where most people play by some basic rules. Unfortunately it is not always the case. 

I know some of what I just said might be controversial, but I want to say leaving the option open is definitely going to be useful for some of the users, especially if it is not contradictory with the RFC itself. Nobody is misleading anybody that there is a mirage of absolute privacy and anonymity with ECH.

**Comment by mholt** [comment]
I guess I don't understand why omitting a public name would help you. The censor would then just block handshakes advertising ECH that lack SNI.

**Comment by christaikobo** [comment]
That's why I was interested in masquerading as cloudflare-ech.com.

Also, without compromising other aspects of the feature, the ADDED possible
benefit wouldn't hurt anyone. Yes censor could react and tighten the
control, but it is possible that it won't.

Also there is snooping as well.

**Comment by shawnz** [comment]
> I guess I don't understand why omitting a public name would help you. The censor would then just block handshakes advertising ECH that lack SNI.

If a server wanted to indiscriminately block ECH, why would they care about whether there's a public name present or not? 

GREASE is the mechanism by which you can discourage providers from indiscriminately blocking ECH, and it doesn't require a public name to be present.

**Comment by mholt** [comment]
> That's why I was interested in masquerading as cloudflare-ech.com.

Wouldn't work, because your IP is not allocated to Cloudflare.

> Another factor to consider is snooping.

What about it though? You haven't presented a tangible explanation as to why omitting a public SNI enhances privacy more than other alternatives.

> If a provider wanted to indiscriminately block ECH, why would they care about whether there's a public name present or not?

We're not talking about indiscriminately blocking of ECH, but rather blocking of individual ECH configs. An empty string is just as easy to block as a non empty SNI string.

**Comment by shawnz** [comment]
Sure, but what's the motivation to block that config? It's no different than a server owner having no server name and a GREASEd ECH block, which doesn't seem especially suspicious to me and could be a common configuration in practice (although I don't have the numbers to back that up)

**Comment by christaikobo** [comment]
When I said snooping, I meant the ISP recording all SNI of traffic from and
to a user.

And why does the server IP has to be allocated to cloudflare? Are you
talking about it wouldn't work at all as an ECH setting? Or are you
suggesting that since the server IP doesn't belong to cloudflare, the outer
SNI spoofing is easy to detect?

**Comment by mholt** [comment]
> Sure, but what's the motivation to block that config? It's no different than a server owner having no server name and a GREASEd ECH block, which doesn't seem especially suspicious to me and could be a common configuration in practice (although I don't have the numbers to back that up)

I think that's my point, an empty outer server name isn't really beneficial.

> When I said snooping, I meant the ISP recording all SNI of traffic from and to a user.

So? Why does the public SNI reveal something secret or private to the ISP?

> And why does the server IP has to be allocated to cloudflare?

The second one:

> Or are you suggesting that since the server IP doesn't belong to cloudflare, the outer SNI spoofing is easy to detect?

It is obvious you aren't connecting to a Cloudflare service.

I still am not seeing a compelling reason to completely omit an outer SNI.

**Comment by christaikobo** [comment]
> So? Why does the public SNI reveal something secret or private to the ISP?

I thought this is the whole point of privacy? There are certain things that I DO NOT want to share: The service I am hosting when I am the server, the service I am visiting when I am the user. By extension then the whole idea of ECH is useless, because Why does the SNI reveal something secret or private to the ISP? I have said earlier that for small organization and individual, hiding private SNI with public SNI is pretty much useless, because you are revealing an SNI anyway and if you only have one domain, it doesn't really matter if you are revealing "a.domain.com" or "b.domain.com".

> It is obvious you aren't connecting to a Cloudflare service.

I have been dealing with the world's largest and most powerful censorship system for a decade now. I am not an expert on the technical side of things, far from it. But I do have a lot of practical experience, and I can say with 100% confidence that large scale censorship system will not deploy the semi targeted investigation you have alluded to, it is difficult to automate, error prone, leads to significant false positive, and most important of all, is an endless game of whack-a-mole. Who says I HAVE TO masquerade as `cloudflare-ech.com`? There are probably hundreds of different CDNs around the world. Or maybe I don't need to masquerade as a CDN at all, just switch my blocked domain with another innocent obscure domain. 

The point is, your idea of the masquerade being too obvious is not reflective of the real situation at all. I am not implying that caddy with ECH enabled can defeat THE censorship system I just mentioned, I am merely using it as an example to illustrate the challenges these kind of censorship systems face, challenges that their adversaries can exploit. 

Maybe you are not seeing the compelling reason because you haven't been in our shoes. I want to reiterate that if it is potentially useful to a not-insignificant number of users, and it is conformant to the RFC, there is no reason to not provide at least the option for the users.

**Comment by shawnz** [comment]
> I think that's my point, an empty outer server name isn't really beneficial.

Well, to be clear I'm not necessarily convinced of the benefits of allowing no outer SNI either, I'm just saying that I don't think there's any incentive for a provider to block that config like you suggested earlier.

But I could think of some niche benefits of allowing it. If the provider blocks you by IP address, you could just get a new IP for your instance without having to change your domain name (which could invalidate links and bookmarks, require software to be redeployed, etc). But if you spoof the outer SNI, that provides opportunities for easy detection that something suspicious is going on. So having no outer SNI might really be the best option for some people in this situation.

**Comment by mholt** [comment]
@christaikobo 

Sorry if I seem dismissive -- I want to be clear that I value your feedback and experience. I'm trying to _understand_ it, and for me, that understanding comes by way of scrutiny. It's how I challenge my current view of the topic. So please forgive me if I break it down, I really am just trying to make sure I am on the same page, since I don't believe there's much published literature on this topic yet. (I haven't found it if there is.) So I feel like we're forging ahead here.

> I have said earlier that for small organization and individual, hiding private SNI with public SNI is pretty much useless, because you are revealing an SNI anyway and if you only have one domain, it doesn't really matter if you are revealing "a.domain.com" or "b.domain.com".

If the outer SNI is not that of the services in the anonymity set, then what is the harm in it?

Also, can you help me understand why "if you only have one domain" is a relevant constraint? I don't think there's any reason you have to only have 1 domain. Do you mean _registered domain_?Even if you do only have one of those, maybe for cost reasons, you could serve the actual service at `service.example.com` but have the public name be `example.com` -- something generic. Or maybe `ech.example.com`. What does that reveal that can't already be inferred from casual inspection (e.g. IP address etc)?

> there is no reason to not provide at least the option for the users.

I don't _think_ it will be technically challenging or development-intensive to provide the option (I'll know better a little further along), but the only reason I'm opposed to just doing this without a really compelling reason is because I'm worried about introducing security/privacy _theater_ into the application. If people think they have more privacy without a public name, and that's not provably true, then that's harmful because it makes the protocol and the whole ecosystem more brittle. Clients will fail to connect entirely -- or worse, resort to insecure methods of connecting as a last resort -- without a public SNI.

Currently, my understanding is that:

- The public name is technically orthogonal to privacy relevant to ECH (according to the scope outlined in the spec).
- An empty public name is technically/functionally equivalent to a public name for which you are not the authoritative server: you cannot get a certificate for it, so retries aren't available.
- An empty public name is no less revealing than a non-empty public name, assuming the non-empty public name is generic (i.e. not related to the services being hidden).

@shawnz 

> I'm just saying that I don't think there's any incentive for a provider to block that config like you suggested earlier.

we've seen things like this before, a "mass casualty" event as it were, where an entire IP subnet will be blocked even though only a specific site or two was being targeted. The site changed IP addresses often enough that the powers that be deemed it better to just block the whole range (temporarily, I think). ECH raises the costs of such censorship, but it's not always too high.

> But I could think of some niche benefits of allowing it. If the provider blocks you by IP address, you could just get a new IP for your instance without having to change your domain name (which could invalidate links and bookmarks, require software to be redeployed, etc). But if you spoof the outer SNI, that provides opportunities for easy detection that something suspicious is going on. So having no outer SNI might really be the best option for some people in this situation.

Hmm, well, assuming you can "just get a new IP" is quite an assumption, as most people don't have control over this, and if they do, it's not cheap. (Residential IPs = almost never; cloud servers = you can destroy and create instances, but this costs; having an IP range allocated to you = usually even more expensive).

Even if the outer SNI is spoofed or missing, ECH is advertised in the handshake, so snoops can always tell that something "suspicious" is going on... (And I want to emphasize that use of privacy features does not imply suspicious activity or content -- for the same reason we must avoid selective encryption, i.e. encrypting only sensitive data -- all connections are suspicious, all content is sensitive. That's the reality.)

@Helanzy

Interesting; so, you're suggesting that some residential ISPs block all TLS handshakes with a non-empty SNI? That seems odd, since if they want to block all HTTP, they should be blocking based on ALPN and/or port, not SNI. Are you sure that's correct?

> Is there any information in ECH's ClientHello to makes it different with existing TLS connections and thus can be detected?

The `encrypted_client_hello` extension (65037) is advertised.

---

Thank you, everyone, for the discussion. I think this is really good for development, and helpful for my understanding, since I don't have your perspectives on things.

**Comment by shawnz** [comment]
> Even if the outer SNI is spoofed or missing, ECH is advertised in the handshake, so snoops can always tell that something "suspicious" is going on...

Sure, but again that's what GREASE is for. Everyone should have GREASEd ECH in the handshake even if they're not using it, to mitigate that possibility

**Comment by christaikobo** [comment]
> Sorry if I seem dismissive -- I want to be clear that I value your feedback and experience. I'm trying to _understand_ it, and for me, that understanding comes by way of scrutiny. It's how I challenge my current view of the topic. So please forgive me if I break it down, I really am just trying to make sure I am on the same page, since I don't believe there's much published literature on this topic yet. (I haven't found it if there is.) So I feel like we're forging ahead here.

Sorry I was a bit impatient as I was a bit frustrated, not being able to get my perspective across.

> > I have said earlier that for small organization and individual, hiding private SNI with public SNI is pretty much useless, because you are revealing an SNI anyway and if you only have one domain, it doesn't really matter if you are revealing "a.domain.com" or "b.domain.com".
> 
> If the outer SNI is not that of the services in the anonymity set, then what is the harm in it?

The restriction that residential user shouldn't host anything the other user mentioned applies as well, SNI is an indication of the existence of HTTPS server. I didn't want to talk about it because it is quite niche.

But more importantly, the thing about privacy is sometimes not always about immediate and tangible harm, which is unlike security. One of the major ISP in my country which has over a billion users used offer this service, in which you could print out all the website you have visited this billing period, that they logged via SNI. It's thousands of entries if not 10s of thousands of entries. And I don't like that at all. I am going to make an analogy: what harm does it have if some random person sees me naked? But I certainly don't want to be seen naked, even if no harm comes my way. 

> Also, can you help me understand why "if you only have one domain" is a relevant constraint? I don't think there's any reason you have to only have 1 domain. Do you mean _registered domain_?Even if you do only have one of those, maybe for cost reasons, you could serve the actual service at `service.example.com` but have the public name be `example.com` -- something generic. Or maybe `ech.example.com`. What does that reveal that can't already be inferred from casual inspection (e.g. IP address etc)?

The thing is a lot of these automated system will never do "casual inspection", they only do automated logging and/or blocking based off certain predefined rules. And SNI is a VERY vital part of these automated systems. They don't have the resources to "casually inspect" every user. Nor would they want to deploy such a unreliable method which could cause massive side effects on a large scale, which I explained in the last comment. 

About "if I only have one domain", if I have 10 domains, and I use `ech.domain-a.com`, I can at least hide the other 9 domains. Again, automated systems will not target me individually and "casually inspect" me, that is just not feasible.

**Comment by mholt** [comment]
> Sure, but again that's what GREASE is for. Everyone should have GREASEd ECH in the handshake even if they're not using it, to mitigate that possibility

Good point, but I also don't think I will reasonably expect that all major clients will implement GREASE... maybe, and it'd be cool if basically every TLS handshake included it. But I dunno, we'll see.

> But more importantly, the thing about privacy is sometimes not always about immediate and tangible harm, which is unlike security. One of the major ISP in my country which has over a billion users used offer this service, in which you could print out all the website you have visited this billing period, that they logged via SNI. It's thousands of entries if not 10s of thousands of entries. And I don't like that at all. I am going to make an analogy: what harm does it have if some random person sees me naked? But I certainly don't want to be seen naked, even if no harm comes my way.

And to clarify, I totally agree: there is a reason we wear clothes... nobody is disagreeing with that. We're just trying to figure out whether it's more private to wear a white shirt than a blue shirt or a graphic tee, etc.

> The thing is a lot of these automated system will never do "casual inspection", they only do automated logging and/or blocking based off certain predefined rules. And SNI is a VERY vital part of these automated systems. 

Right, but this doesn't explain why no SNI is better than an arbitrary non-empty one.

---

I feel like the argument for an empty outer SNI is still beating around the bush. _What's the actual privacy benefit that is so substantial it is worth making client connections brittle?_ So far, the claim remains unsubstantiated. What am I missing?

**Comment by Gunni** [comment]
Having a default of outer SNI enabled with a domain already managed by caddy makes sense, it's the most compatible way.

I just think having the _ability_ to have no outer SNI makes sense, even if it isn't the default. Just like being able to configure what TLS version are allowed, but TLS 1.2 vs TLS 1.3 have no substantial privacy benefits, right?

**Comment by mholt** [comment]
@Gunni Gotcha -- although, to nit-pick, I disagree with:

>  Just like being able to configure what TLS version are allowed, but TLS 1.2 (as configured by caddy by default) vs TLS 1.3 have no substantial privacy benefits, right?

TLS 1.3 _does_ add provable privacy benefits (post-quantum cryptography, for example), and at least at the time, TLS 1.2 was still required for a vast majority of clients to even be compatible.

> I just think having the ability to have no outer SNI makes sense, even if it isn't the default.

To re-iterate what I said above, I hesitate to add this option without provable benefits (I added to my own quote in bold):

> I'm opposed to just doing this without a really compelling reason is because I'm worried about introducing security/privacy theater into the application. If people think they have more privacy without a public name, and that's not provably true, then that's harmful because it makes the protocol and the whole ecosystem more brittle, **[not to mention making people believe they have more privacy when they actually may not]**. Clients will fail to connect entirely -- or worse, resort to insecure methods of connecting as a last resort -- without a public SNI.

**Comment by christaikobo** [comment]
> > I'm opposed to just doing this without a really compelling reason is because I'm worried about introducing security/privacy theater into the application. If people think they have more privacy without a public name, and that's not provably true, then that's harmful because it makes the protocol and the whole ecosystem more brittle, **[not to mention making people believe they have more privacy when they actually may not]**. Clients will fail to connect entirely -- or worse, resort to insecure methods of connecting as a last resort -- without a public SNI.

How about putting the option in an addon? Anyone who uses default binary or docker won't have to option.

And anyone who actively compile the addon into caddy should be aware what they are doing.

**Comment by mholt** [comment]
> Per my understanding, for ClientHelloOuter to play the role to securely provide updated public key on decrypt failure, an encrypted TLS connection should be established using outer SNI. So that means a valid certificate shall be provided for outer SNI, which means we shall own the provided outer SNI domain, right?

Specifically, the server needs to be authoritative for the outer domain, yes.

But that should be no surprise. For example, of course a Cloudflare IP is going to own `cloudflare-ech.com`. Or you might own `alsdkflksjdflksjdf.com` at your residential IP. 🤷‍

> Then, we cannot say it's an arbitrary non-empty domain.

Sure, but what I meant is that it doesn't have to be related to the services being hidden. It can be anything, even a random ASCII string. But yes you're right it can't just be any domain name. It can be random, though.

>  For home users, they may have to use their only domain ech.mydomain or whatever.mydomain as outer SNI.

Huh, I was not aware of this restriction. Can you tell me more?

> Then, the whois information, or regulated registrant for mydomain, or delegated subdomain sub.mydomain to regulated DNS provider, or personal page hosted at etld+1 mydomain may be enough to reveal their identity.

So this is where I would say... don't do that.

(Also, is WHOIS information still publicly available? I wasn't aware it was still published. Just trying a few domains right now and I just get hidden/privacy masks in the output.)

But even so, that all misses the point.

ECH is not meant to hide associations with IP addresses. In fact, associations can be drawn even without domain names. Again, it is an orthogonal concern.

> ECH with outer SNI (with valid certificate) helps with 1 (ALPN is private now) and 2 (accessing ech.mydomain directly will be rejected), but 3 is not trace-proof (especially regarding owning the domain to provide a certificate). With empty outer SNI, 3 is ensured, while 1 is strengthened as there have been reports in my country's ISP using SNI as proof for existence of web service. On the contrary there are more ways to mitigate brittle client connection, e.g. using a private VPN channel to advertise public key in time.

Thanks for elaborating. However I don't think I see things the same way:

> they cannot tell whether it's HTTPS or not.

I _believe_ that should be true, however, be aware that most ISPs / powers-that-be just use port, since the vast majority of HTTPS traffic is on port 443.

> ISP cannot do a replay to see if there's a web service hosted on specific port.

Can you help me understand this better? The spec doesn't mention the word "replay" at _all_ so I would be surprised if this was a design goal.

> ISP cannot trace domain owner's information using outer SNI.

ISP does not need outer SNI to do this. :) IP address is even more informative.

> How about putting the option in an addon?

_Maybe_, but enabling support for a module here would be akin to advocating the capability, I feel like.

**Comment by christaikobo** [comment]
> _Maybe_, but enabling support for a module here would be akin to advocating the capability, I feel like.

First of all, I see your concerns here. Caddy is a very popular tool and TLS is such a core component in its functionality that any change to it will have massive implications to the whole HTTP ecosystem. 

On the other hand, you guys must have statistical metrics about the percentage of users who just use your docker image or download precompile binary from official website/github, as opposed to the percentage of users who use online xcaddy to build custom caddy with addons. 

(Granted not all users use online build tool, but not all use download official docker image or precompiled binary, that kinda evens out)

And I would guess the ratio is quite extreme. My personal thought on this is if a very small percentage of users are willing to go out of their way to experiment on this with awareness of it being potentially breaking, the influence on the whole ecosystem should be limited. What is often harmful to the ecosystem is usually wrong defaults, which is not the case here.

**Comment by mholt** [comment]
@Helanzy 

> The "it can be random domain" statement here is what I don't understand all along. ...  from my understanding we need a certificate signed by acknowledged CAs for outer SNI, to make the handshake success. If we provide a self-signed certificate for random ASCII string, won't the browser just reject to establish the connection?

Yes, so to clarify, it needs to be able to get a valid certificate for the retry mechanism to succeed -- you _may_ use any random domain you own, I don't see why the contents of the domain  are significant? For example, you quoted me saying: "Or you might own `alsdkflksjdflksjdf.com`" -- that is a perfectly fine domain you can use for your outer SNI if you want to. It doesn't have to be, like, `helanzys-services.com`. The domain can be a totally random string.

> but for individuals with only single domain mydomain

And again, I'm stuck on why is this a requirement/restriction? Even if it is, what's the matter with that? Individual services are still masked behind a generic, not-service-specific domain.

> The "replay" detection of web service can work in this way:
>
> 1.   ISP sniffed out SNI service.mydomain:port from HTTPS handshake;
> 2.   ISP visits service.mydomain:port (path is encrypted in HTTPS, so can only view index page).
> 3.   A welcome page is shown -> web service found.

What? I follow step 1... and step 2... but step 3? The point of ECH isn't to hide web services. They are published publicly! The point of ECH is to make it difficult for wiretappers to know which service you are accessing.

@christaikobo Actually, we have no such information. We tried to count that once upon a time, but everyone lost their minds so we stopped.

---

Anyway, at this point, the claim that "ECH without an outer/public name is more private" has not been substantiated/proven, so I will not be implementing that at this time, to prevent potential harm to users and the ecosystem.

(Note I said, "at this time" -- if the claim _is_ proven or demonstrated in a real world scenario, then we can revisit this, and see if it is worth the trade-off of making the system more brittle.)

**Comment by christaikobo** [comment]
> Anyway, at this point, the claim that "ECH without an outer/public name is more private" has not been substantiated/proven, so I will not be implementing that at this time, to prevent potential harm to users and the ecosystem.
> 
> (Note I said, "at this time" -- if the claim _is_ proven or demonstrated in a real world scenario, then we can revisit this, and see if it is worth the trade-off of making the system more brittle.)

I fully understand and respect the decision from the dev. Love all the work you guys have been doing. Although a consensus is not reached this time, caddy team's swift response to industry evolution and open attitude towards user input are much appreciated.

**Comment by gstrauss** [comment]
@mholt wrote:
> I feel like the argument for an empty outer SNI is still beating around the bush. What's the actual privacy benefit that is so substantial it is worth making client connections brittle? So far, the claim remains unsubstantiated. What am I missing?

Perhaps those specifying "no outer SNI" are spitballing a "solution" without fully describing the problem they are trying to solve with that "solution"?

I understood only one attempt to explain "no outer SNI": I think @Helanzy is suggesting that a server configured for ECH which accepts TLS ClientHello without the TLS SNI extension, but with the TLS ECH extension *might not be* detected and blocked by an ISP as a web service.  If that is what @Helanzy is suggesting, then I would like to see a documented example.  That would be suggesting using TLS ECH to work around "all ISP-detected web services getting blocked" -- until the ISP wisened up -- even though that is not directly a privacy benefit of TLS ECH.

---

For others, you can have a wildcard cert for *.my-generic-domain.com as your public server name.  Use random words prepended to .my-generic-domain.com and you can have lots and lots of different public SNI names coming in as requests, enlarging your anonymity set to lots of different public SNI names (though with shared domain suffix).  The inner SNI can be anything else you like.

You can have a self-signed cert, or certs signed with a private CA (or private corporate CA), with lots of different SAN names, including spoofs of names of popular sites.  If the CA signature does not validate or is not trusted by a given client, then the client should not trust the certificate, but is that a concern for you for your ECH public name?  It could be a misconfiguration.  It could be corporate malware producing spoofed certificates for all sites so that the corporate malware could proxy (and view) all HTTPS traffic for the specific corporate clients which trusts the corporate CA.  Other than the situation where your ISP is state-owned and requires an explicit list of state-sanctioned CAs, most other ISPs are not able to enumerate all possible CAs.

---

If you want to have "secret" sites reachable *only* via ECH, and you do not want to expose retry configs, then that might provide an incremental level of privacy.

I am a lighttpd developer.  lighttpd implements TLS ECH with the ability to (attempt) to hide hosts configured in lighttpd.conf to be explicitly accessed via ECH-only.  I implemented the feature of ECH-only hosts in lighttpd as an experiment so that others could try it out and help substantiate whether or not it is measurably useful.

*IF* the ECH public key is distributed out-of-band instead of via public DNS, then an ECH-only service accessed could be "more private" than a typical service accessible via ECH and non-ECH connections.  State-level actors with broad surveillance capabilities can still detect that you are accessing a "secret" service located at a given IP, but might not know specifically which service, and won't not be able to spider the "secret" service without the public key.  (Requiring stronger authentication, which *is not* HTTP Basic Auth, to access the "secret" site is also strongly recommended.)

**Comment by mholt** [comment]
@christaikobo 

> I fully understand and respect the decision from the dev. Love all the work you guys have been doing. Although a consensus is not reached this time, caddy team's swift response to industry evolution and open attitude towards user input are much appreciated.

Thanks for holding a respectful conversation, even though I'm sure it must be frustrating. :)

@Helanzy 

Thanks for the detailed use case. That's very helpful, as I'm otherwise unaware how that works in China.

@gstrauss

> I am a lighttpd developer.

First, thank you for chiming in! Feedback from another web server developer implementing ECH is as about as valuable as it gets for me right now. It's good to know I'm not alone. 😅 Thanks for bringing the focus onto the question at hand and summing things up for me with your knowledge.

So, at this point, my understanding is that:

- It can be desirable to keep which service(s) a host is operating private from parties who are able to observe all incoming and outgoing network activity as it happens (this includes, and is not limited to: origin and destination IPs, number of bytes, protocols used, any unencrypted data, and patterns of encrypted data).

- Having no outer SNI is not necessarily more private with respect to ECH, but if the desire is as stated in the last bullet point, omitting all plaintext server names would be beneficial.

- However, as a certificate is still needed for the inner SNI, a certificate must be procured by the server. This requires DNS records pointing to the machine, which eliminates pure anonymity, OR a DNS challenge record created to prove control over a domain. These records are (generally? always?) created in the clear and at specific times that can be observed on the network.

- Even if A and AAAA records don't exist for the inner domain, an observer of the network can correlate certificate issuance times with certificate transparency (CT) logs to ascertain which domain is being served at the observed host (even if timestamps aren't precise, it vastly narrows the search space).

- Thus, keeping the host's service(s) truly anonymous would require the use of a private/internal CA, that does not publish to CT logs, and whose root would have to be trusted/installed by any/all clients who wish to connect to such services. In other words, as the operator of the private service, you would have to: 1) distribute ECHConfigs in such a way that no adversary could discover, 2) create no public DNS records, 3) obtain no publicly-trusted certificate, 4) operate your own internal/private CA, and 5) ensure the CA's root is trusted by all clients of your service.

Clearly, this falls well beyond the scope of ECH, although ECH would be a crucial component of such a setup. This is the ONLY case I can think of where omitting an outer SNI would provide the privacy benefits being requested (I think); it would essentially eliminate all public traces of the service(s).

(Let me know if I have misunderstood or misrepresented anything!)

**Comment by Gunni** [comment]
You may be forgetting Wild Card DNS records with wild card certificates can completely mask which domain names are in use.

**Comment by gstrauss** [comment]
@mholt, I think you have expressed well that TLS ECH can be *part* of solutions making the web more private, but that there are many, many more details.  Anyone looking at TLS ECH as "the solution" to a problem probably ought to first focus on VPNs, proxies, and other anonymity services before considering TLS ECH.

The Caddy feature for auto-procurement of certificates may leave fingerprints which make hiding the existence of TLS ECH-only hosts more difficult, ... if ECH-only services are even something that can be reasonably achieved (TBD).

---

Regarding TLS ECH implementation, I have not looked into the recently added Go support.  Still, I want to highlight a potential privacy knob.  There is a random byte in the ECHConfig which can help optimize selection of the key on the server used to decrypt the TLS ECH inner, and this feature is desirable to large CDNs.  The RFC draft recognizes that this byte is a possible tracking vector.  The defo.ie patches for OpenSSL include an option for ECHTrialDecrypt to enable trial decryption using a list of keys on the server.  Whether or not to use the ECHConfig random byte -- which should be changed when ECHConfig keys are rotated -- is something to consider, as is whether or not to avoid using random byte (by always pinning to same value) and to force trial decryption to find matching keys.

lighttpd TLS ECH doc includes some notes on using OpenSSL and BoringSSL clients to generate ECHConfig:
https://wiki.lighttpd.net/TLS_ECH
OpenSSL and BoringSSL currently have slight differences in the precise format (wrapping) the generated ECHConfig.
The doc also includes an example of how to test TLS ECH using the cURL client and using private ECHConfig which are not published in DNS.

ECHConfig key rotation should be frequent (at least multiple times per day) and that adds work for servers, including Caddy, to retrieve and rotate keys, including possibly supporting recently-rotated keys for a short period of overlap.

Additional reference: https://github.com/defo-project 
Stephen Farrell (@sftcd) has been one of the primary implementers developing and submitting patches with TLS ECH support to various projects.  He is very approachable if you have questions about TLS ECH.

**Comment by sftcd** [comment]
Happy to help out with ECH here if I can - just ping if there's something I can help with.

IIRC the next golang release was supposed to include the ECH server code, so it should be there now or soon, and our experience is that it's not that hard to ECH-enable a server once you figure out where to put the few new ECH  API calls. (But still happy to help with that if help's useful.)

**Comment by mholt** [comment]
@Gunni 

> You may be forgetting Wild Card DNS records with wild card certificates can completely mask which domain names are in use.

I thought about those, but they would still be subject to the same issuance-time correlation side-channel of any public certificates.

@gstrauss 

Thanks, really great information there!

>  There is a random byte in the ECHConfig which can help optimize selection of the key on the server used to decrypt the TLS ECH inner, and this feature is desirable to large CDNs. The RFC draft recognizes that this byte is a possible tracking vector.

Are you referring to the `config_id`?

> ECHConfig key rotation should be frequent (at least multiple times per day) and that adds work for servers, including Caddy, to retrieve and rotate keys, including possibly supporting recently-rotated keys for a short period of overlap.

That's the plan -- although, I don't know about multiple times _per day_ 😅 I know the spec recommends "frequently" rotating keys, but we can probably manage either way.

@sftcd Nice to meet you, thanks for popping in!

> IIRC the next golang release was supposed to include the ECH server code, so it should be there now or soon, and our experience is that it's not that hard to ECH-enable a server once you figure out where to put the few new ECH API calls. 

Indeed, Go 1.24 supports ECH on the server-side, and its API is very simple. Basically just set [this field](https://pkg.go.dev/crypto/tls#Config.EncryptedClientHelloKeys) in a `tls.Config`.

I'm currently designing the config API for enabling ECH ub Caddy, but then I'll surely have questions as I come to more ECH logistics, such as key rotation. I really appreciate your willingness to chime in :+1:

I'll also hopefully have a WIP PR to share in the next week or two.

**Comment by gstrauss** [comment]
> > There is a random byte in the ECHConfig which can help optimize selection of the key on the server used to decrypt the TLS ECH inner, and this feature is desirable to large CDNs. The RFC draft recognizes that this byte is a possible tracking vector.
> 
> Are you referring to the `config_id`?

Yes.

**Comment by Gunni** [comment]
@mholt

> I thought about those, but they would still be subject to the same issuance-time correlation side-channel of any public certificates.

Yes but you might f.ex get a wildcard cert for example.com, and then only allow specific subdomains to connect, then `close` the connection for others.

**Comment by gdy666** [comment]
In my view, the ECH RFC standard is one aspect, while client-side support (such as in browsers) is another. Therefore, the claim that ECH can hide external domain names or use third-party domain names that do not belong to it is unrealistic. Server-side implementation must not only adhere to the ECH specification but also consider the actual compatibility of clients (browsers); otherwise, it is merely empty talk.

**Comment by mholt** [comment]
Thanks @sftcd  -- great info!

I'm making progress on this, but recently hit a snag... the crypto/tls package expects us to hard-code the ECH keys into a TLS server config, rather than provide a callback function like it lets us do for certificates. That means the keys can't change during the TLS server lifetime, meaning they can't be rotated.

I hesitate to proceed with a design that doesn't support key rotation, because then if it ever is possible in the future, I may have to refactor a lot of code. (Or not; but it's uncertain.)

I might be totally doing something wrong or overlooking something obvious, but I opened an issue upstream just in case: https://github.com/golang/go/issues/71920

**Comment by gstrauss** [comment]
@mholt, I do not know enough of the Go language to comment on the bug report, but want to highlight the importance of having the ability to support multiple ECH keys for trial decryption.  Whether multiple keys are due to key rotation of the same host, or if multiple keys are due to multiple different hosts, sometimes trial decryption is necessary to decrypt the TLS ECH inner, and you will not be able to read SNI from the TLS ECH inner until *after* TLS ECH inner has been decrypted, which is *after* (potentially multiple different) decryption keys have been tried.  Does Go provide a mechanism to configure a list of multiple decryption keys?  Your bug report notes that Go does not provide server side support for dynamic reconfiguration of TLS ECH keys, and if what I am trying to describe is another limitation of the Go TLS ECH server-side implementation, it would be nice to provide that feedback so that the Go devs can suggest a solution to both issues.

**Comment by mholt** [comment]
@gstrauss Thanks -- Go does allow configuring multiple keys: https://pkg.go.dev/crypto/tls#Config.EncryptedClientHelloKeys

The 'EncryptedClientHelloKeys' field lets us specify multiple, which we will be doing in Caddy.

**Comment by gstrauss** [comment]
https://pkg.go.dev/crypto/tls#Config.EncryptedClientHelloKeys
```
	// If EncryptedClientHelloKeys is set, MinVersion, if set, must be
	// VersionTLS13.
```

Does that mean that servers on top of Go will require TLSv1.3 as minimum protocol -- and will not be able to support TLSv1.2 -- or does it mean that TLS ECH is available only with TLSv1.3, as TLS ECH is an extension which requires TLSv1.3?

**Comment by sftcd** [comment]
> I might be totally doing something wrong or overlooking something obvious, but I opened an issue upstream just in case: [golang/go#71920](https://github.com/golang/go/issues/71920)

The bug report looks valid to me (modulo my relative golang ignorance:-). In 'C'/OpenSSL you have the difference between the `SSL_CTX` context structure and the per-connection `SSL` structure so can handle ECH key rotation in the `SSL_CTX`, but I don't see a similar thing at https://pkg.go.dev/crypto/tls#Config.EncryptedClientHelloKeys  In any case, handling ECH key rotation similarly to TLS server private key/cert re-load seems like a good idea, esp given that cloudflare are rotating ECH keys hourly, its not unlikely others will emulate that.

As an aside, that https://pkg.go.dev/crypto/tls#Config thing looks like it might grow up to be like some of the internal OpenSSL structures (i.e. many too many fields) - avoiding that might be a good thing, were it possible:-)

And just for completeness - I very much doubt you'd want to, but haproxy has yet another way to handle key rotation for TLS server keys (and, in my PoC integration ECH keys) - you connect to a `stats` port and use `socat` to load in new private/public values. IIUC their logic in doing that is they don't want to depend on reading the disk any more once haproxy has started (so they continue in the face of some errors).

**Comment by mholt** [comment]
ECH support has been merged and will soon be released with 2.10-beta.1. (We'll have many beta versions for this cycle as I continue to work through the backlog of PRs.)

It's more code than I expected (1500+ LoC), but I tested to make sure at least a few configurations worked properly. Everyone please give it a try!

Note that key rotation is deferred until after Go 1.25 is released, as it's not exactly feasible with Go 1.24.

If you encounter any bugs with it (you surely will), or want to start a new in-depth discussion, please open a new issue for those. Feel free to continue replying here with any smaller questions or comments you may have though.

**Comment by sftcd** [comment]
good stuff - I might try do a few tests, but what's the timeframe for those to be most useful? (i.e. how long is a "cycle")

**Comment by mholt** [comment]
By cycle I just mean the 2.10 release in this case. No specific timeframe yet. 👍
